crates/sui-core/src/authority/authority_notify_read.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::AuthorityStore;
+use crate::checkpoints::causal_order_effects::EffectsStore;
+use async_trait::async_trait;
+use either::Either;
+use futures::future::join_all;
+use parking_lot::Mutex;
+use parking_lot::MutexGuard;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::future::Future;
+use std::hash::{Hash, Hasher};
+use std::mem;
+use std::pin::Pin;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use sui_types::base_types::TransactionDigest;
+use sui_types::error::SuiResult;
+use sui_types::messages::TransactionEffects;
+use tokio::sync::oneshot;
+
+#[allow(dead_code)]
+#[async_trait]
+pub trait EffectsNotifyRead {
+    /// This method reads transaction effects from database.
+    /// If effects are not available immediately, the method blocks until they are persisted
+    /// in the database.
+    ///
+    /// This method **does not** schedule transactions for execution - it is responsibility of the caller
+    /// to schedule transactions for execution before calling this method.
+    async fn notify_read(
+        &self,
+        digests: Vec<TransactionDigest>,
+    ) -> SuiResult<Vec<TransactionEffects>>;
+}
+
+type Registrations<V> = Vec<oneshot::Sender<V>>;
+
+pub(crate) struct NotifyRead<K, V> {
+    pending: Vec<Mutex<HashMap<K, Registrations<V>>>>,
```

Comment:
Why have a Vec of Registrations (which each contain a oneshot) instead of just using a broadcast channel?

Reply:
I am not entirely sure if broadcast channel gives all the guarantees that I want here. I guess I can take a look, but not sure if this would solve any practical problem?